packages/webdriverio/src/utils/index.ts

```diff
@@ -230,8 +233,224 @@ export function isStaleElementError (err: Error) {
     )
 }
 
+/**
+ * handle promise result (resolved or rejected promises)
+ * @param handle browsing context
+ * @param shadowRootManager instance of ShadowRootManager
+ * @param shadowRootId shadow root id that was inspected
+ * @returns a function to handle the result of a shadow root inspection
+ */
+export function elementPromiseHandler <T extends object>(handle: string, shadowRootManager: ShadowRootManager, shadowRootId?: string) {
+    return (el: T | Error) => {
+        const errorString = 'error' in el && typeof el.error === 'string'
+            ? el.error
+            : 'message' in el && typeof el.message === 'string'
+                ? el.message
+                : undefined
+
+        if (errorString) {
+            /**
+             * clear up shadow root if it's not attached to the DOM anymore
+             */
+            if (shadowRootId && errorString.includes('detached shadow root')) {
+                shadowRootManager.deleteShadowRoot(shadowRootId, handle)
+            }
+            return
+        }
+        return el as T
+    }
+}
+
+/**
+ * determine if element is within the scope of another element
+ * @param node an element somewhere located within the shadow DOM tree
+ * @param host the host element (can be a shadow root or the document)
+ * @param handle the browsing context
+ * @returns true if the node is within the scope of the host
+ */
+export async function isWithinElementScope (
+    node: ElementReference,
+    host: WebdriverIO.Element,
+    handle?: string
+): Promise<boolean> {
+    const browser = getBrowserObject(host)
+    const shadowRootManager = getShadowRootManager(browser)
+    const shadowRoots = shadowRootManager.getShadowRootsForContext(handle)
+
+    const [documentFound, ...shadowFinds] = await Promise.all([
+        browser.execute(
+            isDescendent,
+            host as any as HTMLElement,
+            node as any as HTMLElement
+        ).then((wasFound) => wasFound ? host : undefined),
+        ...shadowRoots.map((shadowRootId) => (
+            browser.execute(
```

Comment:
It doesn't need to because it is a parameter within a `Promise.all` which takes a list of promises and resolves them. It is done this way so that all look ups happen in parallel.